### PR TITLE
Refact nl collisions

### PR DESCRIPTION
**Pull request by attacc**
Reduce memory in COLLISIONS for yambo_nl

1) using single precision instead of double-one
2) removing the zero elements (50% of the collisions)

I put tests results here, I have some problem to upload them in the server:
[REPORT-R1-May-15-Wed_16-17.log](https://github.com/yambo-code/yambo/files/15332202/REPORT-R1-May-15-Wed_16-17.log)

### CODE REVIEW

File: src/collisions/COLLISIONS_alloc_and_free.F [hunk 1]
```diff
@@ -191,10 +191,18 @@ subroutine local_alloc_and_free(COLL,COLL_grp)
...
+           if(SP==DP.and.COLLISIONS_load_SP) then
+             YAMBO_ALLOC(COLL(J)%v_c_SP,(COLL(J)%N))
+             COLL(J)%v_c_SP=cZERO
+             mem_size(4)=mem_size(4)+size(COLL(J)%v_c_SP)
+           endif
          endif
        endif
        if (l_clean.and.(l_v1.or.l_all)) then
          YAMBO_FREE(COLL(J)%v_c)
+         if(SP/=DP.and.allocated(COLL(J)%v_c_SP)) then
```
Comment (sangallidavide) [inline src/collisions/COLLISIONS_alloc_and_free.F]: - The allocation is done if (SP==DP.and.COLLISIONS_load_SP) which the allocation is done with the opposite condition. Is this correct? 
- YAMBO_FREE(x) already does a check "if(allocated(x))", no need to put it again.

Comment (attacc) [reply src/collisions/COLLISIONS_alloc_and_free.F]: I just fixed this, and running test again to be sure everything is fine.
I had a small fail in yambo_rt

File: src/collisions/COLLISIONS_compose_nl.F [hunk 2]
```diff
@@ -58,9 +58,13 @@ subroutine COLLISIONS_compose_nl(G_in)
       do ibp=COLL_bands(1),COLL_bands(2)
        do ib=COLL_bands(1),COLL_bands(2)
          !
-         if (HXC_COLL_element(i_coll_mem)%table(ib,ibp,i_kmq)=="t") then
-           RT_Vnl_xc(i_n,i_m,i_k_NL,i_spin)=RT_Vnl_xc(i_n,i_m,i_k_NL,i_spin) &
-&                                               -cI*HXC_COLL_element(i_coll_mem)%v_c(i_comp)*G_in(ib,ibp,i_kmq)
+         if (HXC_COLL_element(i_coll_mem)%table(ib,ibp,i_kmq)=='t') then
+           if(COLLISIONS_load_SP) then
+             COLL=dcmplx(HXC_COLL_element(i_coll_mem)%v_c_SP(i_comp))
```
Comment (sangallidavide) [inline src/collisions/COLLISIONS_compose_nl.F:63]: COLL is of kind SP, why is it defined 
`COLL=dcmplx(X)`  ?
If SP is correct in the definition, it should be
`COLL=cmplx(X,kind=SP)`

File: src/collisions/COLLISIONS_compress.F [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+ ic=0
+ !
+ COLL%table='t'
+ do i_kmq=1,nXkibz
+   do ibp=COLL_bands(1),COLL_bands(2)
+     do ib=COLL_bands(1),COLL_bands(2)
+       !
+       ic=ic+1
+       !
+       if(l_coll_SP)      COLL%v3_c(ib,ibp,i_kmq)=COLL%v_c_SP(ic)
```
Comment (sangallidavide) [inline src/collisions/COLLISIONS_compress.F:49]: Shouldn't this be 
`COLL%v3_c(ib,ibp,i_kmq)=cmplx(COLL%v_c_SP(ic),kind=SP)` ?

File: src/collisions/COLLISIONS_compose_nl.F (hunk 2 above)
Comment (attacc) [reply src/collisions/COLLISIONS_compose_nl.F:63]: Yes, fine, I can change it

**Comment by attacc** [comment]
Let me add that I often calculate collisions with Yambo_rt to save time, and then I use them in Yambo_nl. So many of my calculations were already in single-precision for the collisions just a waste of memory in yambo_nl.

**Pull request closed without merging**